clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java

```diff
@@ -418,13 +418,31 @@ private Long appendWithOffset(long offset, boolean isControlRecord, long timesta
                 appendDefaultRecord(offset, timestamp, key, value, headers);
                 return null;
             } else {
-                return appendLegacyRecord(offset, timestamp, key, value);
+                return appendLegacyRecord(offset, timestamp, key, value, magic);
             }
         } catch (IOException e) {
             throw new KafkaException("I/O exception when writing to the append stream, closing", e);
         }
     }
 
+    /**
+     * Append a record with old format: v0 or v1 only.
+     *
+     * Should only be used for tests only
+     */
+    public Long appendWithOldFormat(long offset, long timestamp, ByteBuffer key, ByteBuffer value, byte magic) {
```

Comment:
Annotate with `VisibleForTesting`?

Reply:
Do we need this? There is already `appendUncheckedWithOffset` which allows you to do something similar. See `LogValidatorTest.recordsWithInvalidInnerMagic`.

Reply:
Ack, I've removed this function and refactored LogValidatorTest.